software/authbox/api.py

```diff
@@ -0,0 +1,171 @@
+  pass
+
+class Config(object):
+  # TODO more than one filename?
+  def __init__(self, filename):
+    self._config = ConfigParser.RawConfigParser()
+    if filename is not None:
+      if not self._config.read([os.path.expanduser(filename)]):
+        # N.b. if config existed but was invalid, we'd get
+        # MissingSectionHeaderError or so.
+        raise Exception('Nonexistent config')
+
+  def get(self, section, option):
+    # Can raise NoOptionError, NoSectionError
+    value = self._config.get(section, option)
+    # Just an optimization
+    if '{' in value:
+      return recursive_config_lookup(value, self._config.items(section))
+    else:
+      return value
+
+  @classmethod
+  def parse_time(cls, time_string):
+    # Returns seconds
+    units = {
+        's': 1,
+        'm': 60,
+        'h': 3600,
+        'd': 86400,
+    }
+    if not time_string:
+      raise Exception('Empty time_string')
+    elif time_string.isdigit():
+      # Fine, seconds?
+      return int(time_string)
+    elif time_string[-1] in units:
+      try:
+        number = int(time_string[:-1])
+      except ValueError:
+        number = float(time_string[:-1])
+      unit_multiplier = units[time_string[-1]]
+      return int(number * unit_multiplier)
+    else:
+      raise Exception('Unknown time_string format', time_string)
+
+  def get_int_seconds(self, section, option, default):
+    if self._config.has_option(section, option):
+      return self.parse_time(self.get(section, option))
+    else:
+      return self.parse_time(default)
+
+
+class BaseDerivedThread(threading.Thread):
+  def __init__(self, event_queue, config_name):
+    # TODO should they also have numeric ids?
+    thread_name = "%s %s" % (self.__class__.__name__, config_name)
+    super(BaseDerivedThread, self).__init__(name=thread_name)
+    self.daemon = True
+
+    self.event_queue = event_queue
+    self.config_name = config_name
+
+
+class BasePinThread(BaseDerivedThread):
+  def __init__(self, event_queue, config_name, input_pin, output_pin):
+    super(BasePinThread, self).__init__(event_queue, config_name)
+
+    self.input_pin = input_pin
+    self.output_pin = output_pin
+
+    GPIO.setmode(GPIO.BOARD)
+    GPIO.setwarnings(False)  # for reusing pins
+    if self.input_pin:
+      GPIO.setup(self.input_pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
+    if self.output_pin:
+      GPIO.setup(self.output_pin, GPIO.OUT)
+
+
+class BaseDispatcher(object):
+  def __init__(self, config):
+    self.config = config
+    self.event_queue = Queue.Queue()  # unbounded
+
+  def load_config_object(self, name, **kwargs):
+    # N.b. args are from config, kwargs are passed from python
+    options = self.config.get('pins', name).split(':')
+    cls_name = options[0]
+    for c in KNOWN_CLASSES:
+      if c.endswith('.' + cls_name):
+        cls = c
+        break
+    else:
+      raise Exception('Unknown item', name)
+    obj = cls(self.event_queue, *options[1:], **kwargs)
```

Comment:
I'm concerned here about the case where there *are* `KNOWN_CLASSES` but non of them end with `cls_name`...

In that case, no exception is raised, but `cls` is never defined... so line 151 is going to throw an `undefined` exception.